expression/collation.go

```diff
@@ -168,131 +192,283 @@ func deriveCoercibilityForColumn(c *Column) Coercibility {
+					repertoire |= arg.Repertoire()
+					continue
+				} else if (isUnicodeCollation(arg.GetType().Charset) && !isUnicodeCollation(dstCharset)) || (arg.GetType().Charset == charset.CharsetUTF8MB4 && dstCharset == charset.CharsetUTF8) {
+					coercibility, dstCharset, dstCollation = arg.Coercibility(), arg.GetType().Charset, arg.GetType().Collate
+					repertoire |= arg.Repertoire()
+					continue
+				} else if repertoire == ASCII && arg.Repertoire() != ASCII {
+					coercibility, dstCharset, dstCollation = arg.Coercibility(), arg.GetType().Charset, arg.GetType().Collate
+					repertoire |= arg.Repertoire()
+					continue
+				} else if repertoire != ASCII && arg.Repertoire() == ASCII {
+					repertoire |= arg.Repertoire()
+					continue
+				}
+			case coercibility > arg.Coercibility():
+				if repertoire == ASCII || coercibility >= CoercibilitySysconst || isUnicodeCollation(arg.GetType().Charset) {
+					coercibility, dstCharset, dstCollation = arg.Coercibility(), arg.GetType().Charset, arg.GetType().Collate
+					repertoire |= arg.Repertoire()
+					continue
+				}
 			}
+
+			// Cannot apply conversion.
+			repertoire |= arg.Repertoire()
+			coercibility, dstCharset, dstCollation = CoercibilityNone, charset.CharsetBin, charset.CollationBin
+			unknownCS = true
+		} else {
+			// If charset is the same, use lower coercibility, if coercibility is the same and none of them are _bin,
+			// derive to CoercibilityNone and _bin collation.
+			switch {
+			case coercibility == arg.Coercibility():
+				if dstCollation == arg.GetType().Collate {
+				} else if coercibility == CoercibilityExplicit {
+					return nil
+				} else if isBinCollation(dstCollation) {
+				} else if isBinCollation(arg.GetType().Collate) {
+					coercibility, dstCharset, dstCollation = arg.Coercibility(), arg.GetType().Charset, arg.GetType().Collate
+				} else {
+					coercibility, dstCollation, dstCharset = CoercibilityNone, getBinCollation(arg.GetType().Charset), arg.GetType().Charset
+				}
+			case coercibility > arg.Coercibility():
+				coercibility, dstCharset, dstCollation = arg.Coercibility(), arg.GetType().Charset, arg.GetType().Collate
+			}
+			repertoire |= arg.Repertoire()
 		}
 	}
 
-	return dstCollation, dstCharset, coercibility, true
+	if unknownCS && coercibility != CoercibilityExplicit {
+		return nil
+	}
+
+	return &ExprCollation{
+		Coer:      coercibility,
+		Repe:      repertoire,
+		Charset:   dstCharset,
+		Collation: dstCollation,
+	}
+}
+
+func isUnicodeCollation(ch string) bool {
+	return ch == charset.CharsetUTF8 || ch == charset.CharsetUTF8MB4
+}
+
+func isBinCollation(collate string) bool {
+	return collate == charset.CollationASCII || collate == charset.CollationLatin1 ||
+		collate == charset.CollationUTF8 || collate == charset.CollationUTF8MB4 ||
+		collate == charset.CollationGBKBin
```

Comment:
How about reusing `IsBinCollation` in collate.go?

Reply:
we can not, `IsBinCollation` is used for clustered index format, `gbk_bin`'s sort_key is not the same with its stored data